_drafts/oryx/orchestration/app-plugins.md

```diff
@@ -0,0 +1,135 @@
+
+The plugin exposes the following API:
+
+- `getInjector()` – returns the injector that it has created
+- `createInjector()` – recreates the injector with the configured providers
+
+For example, to get injector and inject something from it:
+
+```ts
+import { InjectionPlugin } from '@spryker-oryx/core';
+
+app.requirePlugin(InjectionPlugin).inject(...)
+```
+
+The plugin will expose the `App` instance as an `AppRef` token, so that the main application reference is available throughout the application:
+
+```ts
+import { AppRef } from "@spryker-oryx/core";
+import { inject } from "@spryker-oryx/di";
+
+class MyService {
+  constructor(private app = inject(AppRef)) {
+    // Use `app` here
+  }
+}
+```
+
+### Resource plugin
+
+This plugin adds support for the application resources, such as images. The ResourcesPlugin lazy loads resources into the application whenever they're needed.
+
+Whenever resources are added to the Oryx application this plugin will be automatically configured and used.
+
+The plugin exposes the following API:
+
+- `getResources()` – gets all registered resources
+- `getGraphicValue()` – Loads graphical resources
+
+For example, to get all resources:
+
+```ts
+import { ResourcePlugin } from "@spryker-oryx/core";
+
+app.requirePlugin(ResourcePlugin).getResources();
+```
+
+### Theme plugin
+
+This plugin adds support for the Application themes. Whenever a theme is added to the Oryx application this plugin will be automatically configured and used.
+
+The plugin exposes the following API:
+
+- Get configured icons (`getIcons`)
+- Get specific icon (`getIcon`)
+- Get configured breakpoints (`getBreakpoints`)
+- Resolve component theme (`resolve`)
+- Normalize styles (`normalizeStyles`)
+- Interpolate media queries (`generateMedia`)
+
+For example, to get all icons:
+
+```ts
+import { ThemePlugin } from "@spryker-oryx/core";
+
+app.requirePlugin(ThemePlugin).getIcons();
+```
+
+## Plugin Development
+
+You can create custom plugins to influence the behaviour of an Oryx application. The `AppPlugin` is a simple interface that defines:
+
+- `getName()` – It's name string
+- `apply()` – Main life-cycle method
+- `destroy()` – Optional cleanup method
+
+When a plugin is registered on the Oryx application builder it will then be _applied_ by invoking the `AppPlugin.apply()` method with `App` instance as an argument and then it's up to a plugin to decide what it should do and how.
+
+There are two additional plugin life-cycle methods available that are invoked around the main lifecycle of _all_ the plugins:
+
+- `AppPluginBeforeApply` is invoked before the main life-cycle of all the plugins
+- `AppPluginAfterApply` is invoked after the main life-cycle of all the plugins
+
+> **Note** You should never rely on the order of the registration of the plugins to the Oryx application builder and instead use extra plugin lify-cycle methods to establish the order if necessary.
```

Comment:
Why?

Reply:
Because there is no guarantee that consumers will register plugins in their apps in any particular order and for this reason we've created plugin life-cycle methods.

Reply:
Can the order be defined incorrectly and what issues can occur?

Reply:
The point of this message is not to rely on the order so there is no "incorrect" order and no issues will arise because of this.